### PR TITLE
Feature add firebase rules

### LINKED ISSUE #108: Manage user perms on the database
- Users should be able to load any parking
- They should be able to load any review (when loading from the concerned parking)
- They shouldn't be able to load any user. This causes a problem:
> Reviews have a userId as argument. If we can't load the user, how do we display basic info (username and profile picture)?

### PR DESCRIPTION
**Pull request by matttsss**
### What is this PR?
This PR is the first step in linking users to the database

### How?
This was done by adding new rules to the Firebase Cloud Storage and linking the `User` dataclass with the loged in user via their uid.

Closes #108 #109 
### Firebase rules: 
```
rules_version = '2';

service cloud.firestore {
  match /databases/{database}/documents {
		
    match /parkings/{parking} {
    	allow read: if request.auth != null;
    	allow create, write: if 
      	request.auth != null && 
        request.auth.token.firebase.sign_in_provider != 'anonymous';
    }
    
    match /users/{user} {
    	allow create, read, write: if 
        request.auth != null && 
        request.auth.token.firebase.sign_in_provider != 'anonymous' &&
        request.auth.uid == request.resource.data.userId;
    }
    
    match /reviews/{review} {
    	allow read: if request.auth != null;
      allow write, delete: if 
      	request.auth != null && 
        request.auth.token.firebase.sign_in_provider != 'anonymous' && 
        request.auth.uid == request.resource.data.owner;
      allow create: if 
      	request.auth != null && 
        request.auth.token.firebase.sign_in_provider != 'anonymous' && 
        request.auth.uid == request.resource.data.owner;
    }
  }
}
```

### Test coverage: 
image

### CODE REVIEW

File: app/src/main/java/com/github/se/cyrcle/model/user/UserRepository.kt [hunk 1]
```diff
@@ -10,11 +10,11 @@ interface UserRepository {
   fun onSignIn(onSuccess: () -> Unit)
 
   /**
-   * Get a new unique identifier for a parking
+   * Get a a unique identifier for a user
    *
-   * @return a new unique identifier for a parking
+   * @return A UID linked to the user's signed in account
    */
```
Comment (sean-prz) [inline app/src/main/java/com/github/se/cyrcle/model/user/UserRepository.kt:16]: If you think this is a better name you should change it across all repository and viewmodel for consitency, I don't really think which name we use matter, but it should be the same accross the project

File: app/src/main/java/com/github/se/cyrcle/model/user/UserRepositoryFirestore.kt
Comment (Armetiis) [inline app/src/main/java/com/github/se/cyrcle/model/user/UserRepositoryFirestore.kt:1]: I am unsure wether these changes are ideal or not. Wouldn't it be better to let the firebase auth out of the FirestoreRepository and include it in the viewModel. Therefore the function getUid of the viewmodel would directly interact with the auth ? Or have a repo for the auth?
I might be wrong so feel free to explain your point of view

Comment (matttsss) [reply app/src/main/java/com/github/se/cyrcle/model/user/UserRepositoryFirestore.kt:1]: The authentication process is included with Firestore and I feel that for abstraction, it is better to have that in our Firebase Repository

**Review by sean-prz — COMMENTED** [review]
While testing the feature and before diving into the code, I found that indeed an anonymous user could not add spots or review anymore to the database. 
But there is no feedback ! The anonymous user doesn't know that his review can't be added. Could you add a toast or mention in your PR with an issue if this is planned for a future PR .

**Review by sean-prz — COMMENTED** [review]
There is also the fact that a google-signin user must relaunch the app to see his review persistently (leaving and coming back to the card the review does not display). 

Unsure if these are the new rules that creates this bug since I can't revert them to test how it was before. Anyways we should create an issue to not forget this bug

**Comment by matttsss** [comment]
@sean-prz yes, for the moment that error is loged and intensional. UI blocking for anonymous users is planned in #128

**Review by sean-prz — APPROVED** [review]
Thank you for this feature, it really  brings us to a closer to a complete first version.

I would love more details in the description of what you're planning next, since this is the first of a linked series of PR.

File: app/src/main/java/com/github/se/cyrcle/model/user/UserRepositoryFirestore.kt
Comment (Pijukai) [reply app/src/main/java/com/github/se/cyrcle/model/user/UserRepositoryFirestore.kt:1]: Concerning the fact that the view model must manage data and the repository is only for storing it, i feel the same as Armetiis. Now it is not a blocking point for me but can you just detail a bit about the positives points of having it in the FirestoreRepository ?

**Review by Pijukai — APPROVED** [review]
Thank you for this new feature because it is getting us closer to a functional community driven app! I agree with Sean, I am interested in having more details about the content of the future PRs in description. Good Job!

File: app/src/main/java/com/github/se/cyrcle/model/user/UserRepositoryFirestore.kt
Comment (matttsss) [reply app/src/main/java/com/github/se/cyrcle/model/user/UserRepositoryFirestore.kt:1]: Sure, they do store data, however, they also provide services. In our case both are entagled within Firebase. So it makes sense to me to keep it "bundled" together within the repository. Now I do agree that this is not optimal, and I could well imagine us creating an authenticator class to help separate these two sometime in the future.

**Pull request merged**